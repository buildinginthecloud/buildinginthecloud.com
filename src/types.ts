import type { StackProps } from 'aws-cdk-lib';

/**
 * Configuration properties for the MailRelay stack
 */
export interface MailRelayProps extends StackProps {
  /**
   * The domain name for which the mail relay is being configured.
   * This will be used to create the Route53 hosted zone and configure mail records.
   * @default 'buildinginthecloud.com'
   */
  readonly domainName?: string;

  /**
   * The target domain name for HTTPS redirect.
   * All HTTP/HTTPS traffic to the domain will be redirected to this target domain.
   * @default 'yvovanzee.nl'
   */
  readonly targetDomainName?: string;
}

/**
 * Configuration properties for the DomainRedirect stack
 */
export interface DomainRedirectProps extends StackProps {
  /**
   * The source domain name that will redirect to the target domain.
   * @default 'buildinginthecloud.com'
   */
  readonly sourceDomain?: string;

  /**
   * The target domain name where traffic will be redirected.
   * @default 'yvovanzee.nl'
   */
  readonly targetDomain?: string;

  /**
   * Whether to preserve the URL path in the redirect.
   * @default true
   */
  readonly preservePath?: boolean;

  /**
   * Whether to force HTTPS in the redirect.
   * @default true
   */
  readonly forceHttps?: boolean;

  /**
   * HTTP redirect code to use.
   * @default 301
   */
  readonly redirectCode?: number;
}

/**
 * Redirect configuration interface
 */
export interface RedirectConfig {
  /** Source domain (e.g., buildinginthecloud.com) */
  readonly sourceDomain: string;

  /** Target domain (e.g., yvovanzee.nl) */
  readonly targetDomain: string;

  /** Whether to preserve URL path in redirect */
  readonly preservePath: boolean;

  /** Whether to force HTTPS */
  readonly forceHttps: boolean;

  /** HTTP redirect status code */
  readonly redirectCode: number;
}

/**
 * Mail record configuration constants
 */
export const MAIL_CONFIG = {
  /** Time-to-live for DNS records in minutes */
  TTL: 1800,

  /** SPF record value for iCloud mail */
  SPF_RECORD: 'v=spf1 include:icloud.com ~all',

  /** MX record priorities and hostnames */
  MX_RECORDS: [
    { priority: 10, hostName: 'mx01.mail.icloud.com' },
    { priority: 20, hostName: 'mx02.mail.icloud.com' },
  ] as const,

  TXT_RECORD: 'apple-domain=Qc2qLE9vanUyATjL',
} as const;
