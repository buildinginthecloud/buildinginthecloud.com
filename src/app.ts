// ~~ Generated by projen
/* eslint-disable */
import { App, AppProps, Stack, StackProps } from 'aws-cdk-lib';

/**
 * PipelineAppProps is an extension of AppProps, which is part of the AWS CDK core.
 * It includes optional functions to provide AWS Stacks for different stages.
 *
 * Use these functions to instantiate your application stacks with the parameters for
 * each stage
 */
export interface PipelineAppProps extends AppProps {
  /** This function will be used to generate a production stack. */
  provideProductionStack: (app: App, stackId: string, props: PipelineAppStackProps) => Stack;

}

/**
 * PipelineAppStackProps is an extension of StackProps, which is part of the AWS CDK core.
 * It includes an additional property to specify the stage name.
 */
export interface PipelineAppStackProps extends StackProps {
  stageName: string;
}

/**
 * The PipelineApp class extends the App class from AWS CDK and overrides the constructor to support
 * different stages of the application (development, production, personal, feature) by invoking the provided
 * stack-providing functions from the props.
 */
export class PipelineApp extends App {
  constructor(props: PipelineAppProps) {
    super(props);

    // If a function is provided for creating a production stack, it is called with necessary arguments.
    if (props.provideProductionStack) {
      props.provideProductionStack(this, 'BuildingintheCloud-production', { env: {"account":"730335247138","region":"eu-west-1"}, stackName: 'BuildingintheCloud-production', stageName: 'production' });
    }


  }
}
