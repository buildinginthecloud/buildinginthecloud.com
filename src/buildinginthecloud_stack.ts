import { Stack, StackProps, aws_route53, aws_route53_patterns } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface BuildingInTheCloudStackProps extends StackProps {
  hostedZoneName: string;
  targetDomainName: string;
}

export class BuildingInTheCloudStack extends Stack {
  constructor(scope: Construct, id: string, props: BuildingInTheCloudStackProps) {
    super(scope, id, props);

    const HostedZone = new aws_route53.HostedZone(this, 'HostedZone', {
      zoneName: props.hostedZoneName,
    });

    new aws_route53_patterns.HttpsRedirect(this, 'Redirect', {
      recordNames: [props.hostedZoneName],
      targetDomain: props.targetDomainName,
      zone: HostedZone,
    });
  }
}