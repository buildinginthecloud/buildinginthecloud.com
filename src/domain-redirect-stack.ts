import { Stack, Duration, RemovalPolicy, CfnOutput } from 'aws-cdk-lib';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as route53Patterns from 'aws-cdk-lib/aws-route53-patterns';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { type Construct } from 'constructs';
import type { DomainRedirectProps, RedirectConfig } from './types';

/**
 * A CDK Stack that configures domain redirect from buildinginthecloud.com to yvovanzee.nl.
 * This stack creates:
 * - S3 bucket with static website hosting for redirect rules
 * - CloudFront distribution for HTTPS and global edge locations
 * - Route53 DNS records for domain routing
 * - SSL certificate via ACM for HTTPS support
 *
 * @example
 * ```typescript
 * new DomainRedirectStack(app, 'DomainRedirectStack', {
 *   sourceDomain: 'buildinginthecloud.com',
 *   targetDomain: 'yvovanzee.nl'
 * });
 * ```
 */
export class DomainRedirectStack extends Stack {
  /**
   * The Route53 hosted zone for the source domain.
   */
  public readonly hostedZone: route53.IHostedZone;

  /**
   * The HTTPS redirect construct from Route53 patterns.
   */
  public readonly httpsRedirect: route53Patterns.HttpsRedirect;

  /**
   * The redirect configuration used by this stack.
   */
  public readonly redirectConfig: RedirectConfig;

  /**
   * S3 bucket for storing CloudFront access logs.
   */
  public readonly logsBucket: s3.Bucket;

  /**
   * CloudWatch alarms for monitoring redirect functionality.
   */
  public readonly alarms: cloudwatch.Alarm[];

  constructor(scope: Construct, id: string, props: DomainRedirectProps = {}) {
    super(scope, id, props);

    // Create redirect configuration with defaults
    this.redirectConfig = {
      sourceDomain: props.sourceDomain ?? 'buildinginthecloud.com',
      targetDomain: props.targetDomain ?? 'yvovanzee.nl',
      preservePath: props.preservePath ?? true,
      forceHttps: props.forceHttps ?? true,
      redirectCode: props.redirectCode ?? 301,
    };

    // Create or reference Route53 hosted zone
    this.hostedZone = this.getRoute53HostedZone();

    // Create S3 bucket for access logs
    this.logsBucket = this.createLogsBucket();

    // Create HTTPS redirect using Route53 patterns
    this.httpsRedirect = this.createHttpsRedirect();

    // Configure CloudWatch monitoring and alarms (temporarily disabled)
    this.alarms = [];

    // Create stack outputs for monitoring
    this.createStackOutputs();
  }

  /**
   * References the existing Route53 hosted zone for the source domain.
   */
  private getRoute53HostedZone(): route53.IHostedZone {
    // Reference existing hosted zone for source domain
    const hostedZone = route53.HostedZone.fromLookup(this, 'SourceDomainHostedZone', {
      domainName: this.redirectConfig.sourceDomain,
    });

    return hostedZone;
  }

  /**
   * Creates HTTPS redirect using Route53 patterns.
   * This handles S3 bucket, CloudFront distribution, SSL certificate, and DNS records automatically.
   * Requirements: 1.1, 1.2, 2.1, 2.2, 2.3, 2.4
   */
  private createHttpsRedirect(): route53Patterns.HttpsRedirect {
    const httpsRedirect = new route53Patterns.HttpsRedirect(this, 'HttpsRedirect', {
      recordNames: [this.redirectConfig.sourceDomain, `www.${this.redirectConfig.sourceDomain}`],
      targetDomain: this.redirectConfig.targetDomain,
      zone: this.hostedZone,
    });

    return httpsRedirect;
  }

  /**
   * Creates S3 bucket for storing CloudFront access logs.
   * Requirements: 2.4, 3.4
   */
  private createLogsBucket(): s3.Bucket {
    const logsBucketName = `${this.redirectConfig.sourceDomain.replace(/\./g, '-')}-logs-${this.account}`;

    const logsBucket = new s3.Bucket(this, 'LogsBucket', {
      bucketName: logsBucketName,
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      lifecycleRules: [
        {
          id: 'DeleteOldLogs',
          enabled: true,
          expiration: Duration.days(90), // Keep logs for 90 days
          transitions: [
            {
              storageClass: s3.StorageClass.INFREQUENT_ACCESS,
              transitionAfter: Duration.days(30),
            },
            {
              storageClass: s3.StorageClass.GLACIER,
              transitionAfter: Duration.days(60),
            },
          ],
        },
      ],
    });

    // Grant CloudFront permission to write logs
    logsBucket.addToResourcePolicy(
      new iam.PolicyStatement({
        sid: 'AllowCloudFrontLogging',
        effect: iam.Effect.ALLOW,
        principals: [new iam.ServicePrincipal('cloudfront.amazonaws.com')],
        actions: ['s3:PutObject', 's3:GetBucketAcl', 's3:PutBucketAcl'],
        resources: [logsBucket.bucketArn, `${logsBucket.bucketArn}/*`],
      }),
    );

    return logsBucket;
  }

  /**
   * Creates CloudWatch monitoring and alarms for the redirect functionality.
   * Sets up metrics for CloudFront distribution and creates alarms for redirect failures.
   * Requirements: 2.4, 3.4
   * 
   * Note: Temporarily disabled until we determine the correct HttpsRedirect properties
   */
  private createCloudWatchMonitoring(): cloudwatch.Alarm[] {
    const alarms: cloudwatch.Alarm[] = [];

    // TODO: Re-enable monitoring once we determine the correct property names
    // const distribution = this.httpsRedirect.???;

    // TODO: Re-enable monitoring alarms and dashboard

    return alarms;
  }

  /**
   * Creates CloudFormation outputs for monitoring and operational purposes.
   * Requirements: 3.3
   */
  private createStackOutputs(): void {
    // Note: CloudFront distribution outputs temporarily disabled
    // Will be re-enabled once we determine the correct property names

    // Route53 Hosted Zone ID
    new CfnOutput(this, 'HostedZoneId', {
      value: this.hostedZone.hostedZoneId,
      description: 'Route53 Hosted Zone ID for DNS management',
      exportName: `${this.stackName}-HostedZoneId`,
    });

    // Route53 Hosted Zone Name (using existing zone)
    new CfnOutput(this, 'HostedZoneName', {
      value: this.redirectConfig.sourceDomain,
      description: 'Route53 Hosted Zone Name (using existing zone)',
      exportName: `${this.stackName}-HostedZoneName`,
    });

    // Redirect Configuration Summary
    new CfnOutput(this, 'RedirectConfiguration', {
      value: `${this.redirectConfig.sourceDomain} -> ${this.redirectConfig.targetDomain} (301 redirect)`,
      description: 'Redirect configuration summary',
      exportName: `${this.stackName}-RedirectConfiguration`,
    });

    // Deployment Timestamp
    new CfnOutput(this, 'DeploymentTimestamp', {
      value: new Date().toISOString(),
      description: 'Deployment timestamp for tracking',
      exportName: `${this.stackName}-DeploymentTimestamp`,
    });

    // Logs Bucket Name
    new CfnOutput(this, 'LogsBucketName', {
      value: this.logsBucket.bucketName,
      description: 'S3 Bucket name for CloudFront access logs',
      exportName: `${this.stackName}-LogsBucketName`,
    });

    // CloudWatch Dashboard URL
    new CfnOutput(this, 'MonitoringDashboardUrl', {
      value: `https://${this.region}.console.aws.amazon.com/cloudwatch/home?region=${this.region}#dashboards:name=${this.redirectConfig.sourceDomain.replace(/\./g, '-')}-redirect-monitoring`,
      description: 'CloudWatch Dashboard URL for monitoring redirect performance',
      exportName: `${this.stackName}-MonitoringDashboardUrl`,
    });

    // CloudWatch Alarms Summary
    new CfnOutput(this, 'CloudWatchAlarmsCount', {
      value: this.alarms.length.toString(),
      description: 'Number of CloudWatch alarms configured for monitoring',
      exportName: `${this.stackName}-CloudWatchAlarmsCount`,
    });
  }
}
