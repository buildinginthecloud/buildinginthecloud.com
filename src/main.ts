import { Stack, App } from 'aws-cdk-lib';
import { type Construct } from 'constructs';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as route53_patterns from 'aws-cdk-lib/aws-route53-patterns';
import { Duration } from 'aws-cdk-lib';
import type { MailRelayProps } from './types';
import { MAIL_CONFIG } from './types';

/**
 * A CDK Stack that configures mail relay services using Route53 for Apple iCloud mail services.
 * This stack creates:
 * - A Route53 hosted zone for the domain
 * - SPF record for mail verification
 * - DKIM record for domain key validation
 * - MX records for mail routing
 * - HTTPS redirect for web traffic
 * 
 * @example
 * ```typescript
 * new MailRelay(app, 'MailRelayStack', {
 *   domainName: 'example.com',
 *   targetDomainName: 'target.com'
 * });
 * ```
 */
export class MailRelay extends Stack {
  /**
   * The Route53 hosted zone for the domain.
   * This zone contains all the DNS records for mail configuration.
   */
  public readonly hostedZone: route53.PublicHostedZone;

  constructor(scope: Construct, id: string, props: MailRelayProps = {}) {
    super(scope, id, props);

    // Extract domain names with defaults
    const domainName = props.domainName ?? 'buildinginthecloud.com';
    const targetDomainName = props.targetDomainName ?? 'yvovanzee.nl';

    // Create a new public hosted zone for the domain
    this.hostedZone = new route53.PublicHostedZone(this, 'HostedZone', {
      zoneName: domainName,
    });

    // Create a TXT Record for apple mail verification (SPF)
    new route53.TxtRecord(this, 'TXTRecord', {
      zone: this.hostedZone,
      // recordName: `apple-domain=${domainName}`,
      values: [MAIL_CONFIG.TXT_RECORD],
      ttl: Duration.minutes(MAIL_CONFIG.TTL),
    });
    new route53.TxtRecord(this, 'SPFRecord', {
      zone: this.hostedZone,
      recordName: `@`,
      values: [MAIL_CONFIG.SPF_RECORD],
      ttl: Duration.minutes(MAIL_CONFIG.TTL),
    })

    // Create CNAME record for DKIM validation
    new route53.CnameRecord(this, 'DKIMRecord', {
      zone: this.hostedZone,
      recordName: `sig1._domainkey.${domainName}`,
      domainName: `sig1.dkim.${domainName}.at.icloudmailadmin.com`,
      ttl: Duration.minutes(MAIL_CONFIG.TTL),
    });

    // Create MX Record pointing to iCloud mail servers
    new route53.MxRecord(this, 'MXRecord', {
      zone: this.hostedZone,
      recordName: domainName,
      values: [...MAIL_CONFIG.MX_RECORDS],
      ttl: Duration.minutes(MAIL_CONFIG.TTL),
    });

    // Create HTTPS redirect for web traffic
    new route53_patterns.HttpsRedirect(this, 'Redirect', {
      recordNames: [this.hostedZone.zoneName],
      targetDomain: targetDomainName,
      zone: this.hostedZone,
    });
  }
}

// Environment configuration for development
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
} as const;

// Create and configure the CDK app
const app = new App();

new MailRelay(app, 'buildinginthecloud-dev', { env: devEnv });
// new MailRelay(app, 'buildinginthecloud-prod', { env: prodEnv });

app.synth();

