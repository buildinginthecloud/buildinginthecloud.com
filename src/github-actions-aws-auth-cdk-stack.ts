import { GithubActionsIdentityProvider, GithubActionsRole } from 'aws-cdk-github-oidc';
import { CfnOutput, Stack, type StackProps } from 'aws-cdk-lib';
import { ManagedPolicy } from 'aws-cdk-lib/aws-iam';
import type { Construct } from 'constructs';
import { getGitRepoDetails } from '../src/git-helper';

interface GitHubOIDCStackProps extends StackProps {}

export class GitHubOIDCStack extends Stack {
  constructor(scope: Construct, id: string, props: GitHubOIDCStackProps) {
    super(scope, id, props);

    const { gitOwner, repoName } = getGitRepoDetails();

    const provider = GithubActionsIdentityProvider.fromAccount(this, 'GithubProvider');

    const deployRole = new GithubActionsRole(this, 'GitHubDeployRole', {
      provider: provider,
      owner: gitOwner,
      repo: repoName,
      roleName: 'GitHubDeployRole',
      managedPolicies: [ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess')],
    });

    new CfnOutput(this, 'DeployRole', {
      value: deployRole.roleArn,
    });
  }
}