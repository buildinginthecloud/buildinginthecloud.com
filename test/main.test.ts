import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { MailRelay } from '../src/main';
import { MAIL_CONFIG } from '../src/types';

describe('MailRelay Stack', () => {
  let app: App;
  let stack: MailRelay;
  let template: Template;

  beforeEach(() => {
    app = new App();
    stack = new MailRelay(app, 'TestStack', {
      domainName: 'test-domain.com',
      targetDomainName: 'target-domain.com',
    });
    template = Template.fromStack(stack);
  });

  test('creates a Route53 hosted zone', () => {
    template.hasResourceProperties('AWS::Route53::HostedZone', {
      Name: 'test-domain.com.',
    });
  });

  test('creates SPF record', () => {
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'TXT',
      Name: 'apple-domain=test-domain.com.',
      ResourceRecords: [MAIL_CONFIG.SPF_RECORD],
      TTL: String(MAIL_CONFIG.TTL * 60), // TTL in seconds
    });
  });

  test('creates DKIM CNAME record', () => {
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'CNAME',
      Name: 'sig1._domainkey.test-domain.com.',
      ResourceRecords: ['sig1.dkim.test-domain.com.at.icloudmailadmin.com'],
      TTL: String(MAIL_CONFIG.TTL * 60),
    });
  });

  test('creates MX records', () => {
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'MX',
      Name: 'test-domain.com.',
      ResourceRecords: [
        '10 mx01.mail.icloud.com',
        '20 mx02.mail.icloud.com',
      ],
      TTL: String(MAIL_CONFIG.TTL * 60),
    });
  });

  test('creates HTTPS redirect', () => {
    template.resourceCountIs('AWS::CloudFront::Distribution', 1);
    template.hasResourceProperties('AWS::Route53::RecordSet', {
      Type: 'A',
      Name: 'test-domain.com.',
    });
  });

  test('snapshot test', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
