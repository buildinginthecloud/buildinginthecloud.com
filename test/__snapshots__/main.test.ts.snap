// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Mappings": {
    "AWSCloudFrontPartitionHostedZoneIdMap": {
      "aws": {
        "zoneId": "Z2FDTNDATAQYW2",
      },
      "aws-cn": {
        "zoneId": "Z3RFFRIM2A3IF5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "HostedZoneDB99F866": {
      "Properties": {
        "Name": "buildinginthecloud.com.",
      },
      "Type": "AWS::Route53::HostedZone",
    },
    "RedirectRedirectAliasRecordSixc02fee2A734979": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "RedirectRedirectDistributionCFDistribution9866C854",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "buildinginthecloud.com.",
        "Type": "AAAA",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RedirectRedirectAliasRecordc02fee2436FEDB": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "RedirectRedirectDistributionCFDistribution9866C854",
              "DomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::FindInMap": [
              "AWSCloudFrontPartitionHostedZoneIdMap",
              {
                "Ref": "AWS::Partition",
              },
              "zoneId",
            ],
          },
        },
        "HostedZoneId": {
          "Ref": "HostedZoneDB99F866",
        },
        "Name": "buildinginthecloud.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "RedirectRedirectBucketC989E6F1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "WebsiteConfiguration": {
          "RedirectAllRequestsTo": {
            "HostName": "yvovanzee.nl",
            "Protocol": "https",
          },
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "RedirectRedirectCertificateCertificateRequestorFunction6922BC5C": {
      "DependsOn": [
        "RedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyF6E20B96",
        "RedirectRedirectCertificateCertificateRequestorFunctionServiceRole07F91D2B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b073cebcf4d61fb152a30f5a5e57a94df7f980a549fdf1a79a0b18c5750522d8.zip",
        },
        "Handler": "index.certificateRequestHandler",
        "Role": {
          "Fn::GetAtt": [
            "RedirectRedirectCertificateCertificateRequestorFunctionServiceRole07F91D2B",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RedirectRedirectCertificateCertificateRequestorFunctionServiceRole07F91D2B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyF6E20B96": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
                "acm:AddTagsToCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "route53:changeResourceRecordSets",
              "Condition": {
                "ForAllValues:StringEquals": {
                  "route53:ChangeResourceRecordSetsActions": [
                    "UPSERT",
                  ],
                  "route53:ChangeResourceRecordSetsRecordTypes": [
                    "CNAME",
                  ],
                },
                "ForAllValues:StringLike": {
                  "route53:ChangeResourceRecordSetsNormalizedRecordNames": [
                    "*.buildinginthecloud.com",
                    "*.buildinginthecloud.com",
                  ],
                },
              },
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":route53:::hostedzone/",
                    {
                      "Ref": "HostedZoneDB99F866",
                    },
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RedirectRedirectCertificateCertificateRequestorFunctionServiceRoleDefaultPolicyF6E20B96",
        "Roles": [
          {
            "Ref": "RedirectRedirectCertificateCertificateRequestorFunctionServiceRole07F91D2B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "RedirectRedirectCertificateCertificateRequestorResource9F9673DE": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DomainName": "buildinginthecloud.com",
        "HostedZoneId": {
          "Ref": "HostedZoneDB99F866",
        },
        "Region": "us-east-1",
        "ServiceToken": {
          "Fn::GetAtt": [
            "RedirectRedirectCertificateCertificateRequestorFunction6922BC5C",
            "Arn",
          ],
        },
        "SubjectAlternativeNames": [
          "buildinginthecloud.com",
        ],
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "RedirectRedirectDistributionCFDistribution9866C854": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "buildinginthecloud.com",
          ],
          "Comment": "Redirect to yvovanzee.nl from buildinginthecloud.com",
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
            ],
            "CachedMethods": [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": {
              "Cookies": {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Fn::GetAtt": [
                          "RedirectRedirectBucketC989E6F1",
                          "WebsiteURL",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Id": "origin1",
            },
          ],
          "PriceClass": "PriceClass_All",
          "ViewerCertificate": {
            "AcmCertificateArn": {
              "Fn::GetAtt": [
                "RedirectRedirectCertificateCertificateRequestorResource9F9673DE",
                "Arn",
              ],
            },
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
